src/simplify_algebra.cpp

```diff
@@ -412,24 +412,83 @@ struct find_dot_mul
     }
 };
 
+
+
+auto fusable_split()
+{
+    return match::make_basic_pred_matcher([](instruction_ref ins) {
+
+        std::vector<instruction_ref> slices;
+
+        for (const auto& output : ins->outputs())
+        {
+            if (output->name() == "slice")
+            {
+                slices.push_back(output);
+            }
+        }
+
+        if (slices.empty())
+            return false;
+
+        std::vector<instruction_ref> add_instructions;
+        for (const auto& slice : slices)
+        {
+            bool used_by_add = false;
+            for (auto& user : slice->outputs())
+            {
+                if (user->name() == "add")
+                {
+                    used_by_add = true;
+                    add_instructions.push_back(user);
+                }
+            }
+            if (!used_by_add)
+                return false;
+        }        
+
+        bool any_add_followed_by_mul = false;
+        for (const auto& add_ins : add_instructions)
+        {
+            for (auto& user : add_ins->outputs())
+            {
+                if (user->name() == "mul")
+                {
+                    any_add_followed_by_mul = true;
+                    ins->outputs();
+                }
+            }
+        }
```

Comment:
All this is really hard to follow and the intermediate vectors are really unnecessary. This can be done with nested all_of+any_of algorithms:

```cpp
auto fusable_split(const std::string& name)
{
    return match::make_basic_pred_matcher([&](instruction_ref ins) {
        return all_of(ins->outputs(), [&](instruction_ref slice) {
            if(output->name() != "slice")
                return true;
            return any_of(slice->outputs(), [&](instruction_ref x) {
                return x->name() == name;
            });
        });
    });
}
```

You still might need an extra `any_of` check for the case when there is no slices.